### PR TITLE
docs: document errors for a few functions

### PR DESCRIPTION
**Pull request by WhyNotHugo**

### CODE REVIEW

File: zbus/src/connection/builder.rs [hunk 1]
```diff
@@ -77,11 +77,19 @@ assert_impl_all!(Builder<'_>: Send, Sync, Unpin);
 
 impl<'a> Builder<'a> {
     /// Create a builder for the session/user message bus connection.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error variant if parsing the session bus address fails.
```
Comment (zeenix) [inline zbus/src/connection/builder.rs:83]: nit: We should either remove "variant" or explicitly mention the variant name.

File: zbus/src/connection/builder.rs [hunk 2]
```diff
@@ -266,6 +274,10 @@ impl<'a> Builder<'a> {
     /// interfaces available immediately after the connection is established. Typically, this is
     /// exactly what you'd want. Also in contrast to [`zbus::ObjectServer::at`], this method will
     /// replace any previously added interface with the same name at the same path.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if and only if calling `ObjectPath::try_from(path)` fails.
```
Comment (zeenix) [inline zbus/src/connection/builder.rs:280]: `try_from` is an implementation detail, one that's obvious from the generic constraint. I'd write "if `path` is not a valid object path".

**Review by zeenix — CHANGES_REQUESTED** [review]
Let's be consistent and add these for all failing methods here? Also, don't forget the `blocking::connection::Builder` wrapper.

File: zbus/src/connection/builder.rs (hunk 1 above)
Comment (WhyNotHugo) [reply zbus/src/connection/builder.rs:83]: I don't think this is terribly useful, but I've mentioned the variant explicitly anyway.

File: zbus/src/connection/builder.rs (hunk 2 above)
Comment (WhyNotHugo) [reply zbus/src/connection/builder.rs]: Done.

**Comment by WhyNotHugo** [comment]
> Let's be consistent and add these for all failing methods here?

Of course, that's the Rust convention. There's a clippy lint to find all of the missing cases: https://rust-lang.github.io/rust-clippy/master/index.html#/missing_errors_doc

File: zbus/src/connection/builder.rs [hunk 3]
```diff
@@ -77,11 +77,19 @@ assert_impl_all!(Builder<'_>: Send, Sync, Unpin);
 
 impl<'a> Builder<'a> {
     /// Create a builder for the session/user message bus connection.
+    ///
+    /// # Errors
+    ///
+    /// Returns an [`Err`](Result::Err) variant if parsing the session bus address fails.
```
Comment (zeenix) [inline zbus/src/connection/builder.rs:83]: Now I understand what you actually meant by "variant" here and why you disagreed with my comment. I thought you mean the "zbus::Error" variant since Result is obvious to even new Rust devs.

My suggest was to either mention the exact `zbus::Error` variant that can be returned from here or just say "error". AFAIK this is the convention too and is useful.

File: zbus/src/connection/builder.rs [hunk 4]
```diff
@@ -266,6 +274,11 @@ impl<'a> Builder<'a> {
     /// interfaces available immediately after the connection is established. Typically, this is
     /// exactly what you'd want. Also in contrast to [`zbus::ObjectServer::at`], this method will
     /// replace any previously added interface with the same name at the same path.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if path cannot be parsed as a valid object path (i.e.: if
+    /// `ObjectPath::try_from(path)` fails).
```
Comment (zeenix) [inline zbus/src/connection/builder.rs:281]: you still kept the internal detail. Please remove the `(..)` part. It's not needed.

**Review by zeenix — CHANGES_REQUESTED** [review]

**Comment by zeenix** [comment]
> > Let's be consistent and add these for all failing methods here?
> 
> Of course,

Cool. Just to be clear, you mean you're going to update the commit to add them?

> that's the Rust convention. There's a clippy lint to find all of the missing cases: https://rust-lang.github.io/rust-clippy/master/index.html#/missing_errors_doc

Nice! I didn't know about this lint.

File: zbus/src/connection/builder.rs (hunk 3 above)
Comment (WhyNotHugo) [reply zbus/src/connection/builder.rs]: I think that's more useful. Updated.

File: zbus/src/connection/builder.rs (hunk 4 above)
Comment (WhyNotHugo) [reply zbus/src/connection/builder.rs:283]: The part between the parenthesis is where the actually useful information is contained.

I was writing some code that calls this method, and wasn't quite sure when this could return and error (and therefore, how to handle it). I'm passing an object of type `ObjectPath`. A description saying _"Returns an error if path cannot be parsed as a valid object path"_ doesn't say anything useful, because it's not clear at all _when_ this can fail or _why_.

By specifying that this fails when `ObjectPath::try_from(path)` would fail, then it is clear that this call is likely infallible (because `T::try_from(t: T)` is usually infallible), so I can just use `.expect()` assuming that this does not fail.

---

TBH, taking a `TryInto<T>` as input in a function like this is an anti-pattern. It forces the API to have to represent the error even when it is not applicable, even when it doesn't make sense. It would make much more sense to take an `ObjectPath` as input, and require that the caller do the validation/conversion [if necessary] themselves.

However, this isn't about fixing the API; I'm just trying to make the existing API clearer so it's more obvious when errors need to be handled and how.

**Comment by WhyNotHugo** [comment]
> Cool. Just to be clear, you mean you're going to update the commit to add them?

No, that can be done in a follow-up PR (likely several different ones). There are another 153 occurrences of the same information missing, and each occurrence requires careful inspection to understand when they return an error.

File: zbus/src/connection/builder.rs (hunk 4 above)
Comment (MaxVerevkin) [reply zbus/src/connection/builder.rs:283]: I feel similarly about this API (#321)

Comment (zeenix) [reply zbus/src/connection/builder.rs:283]: > because it's not clear at all _when_ this can fail or _why_.

As a user, you don't need to know. If you do, look at the generic constraints. User just needs to know that they need to pass a valid object path. If they don't, they'll get an error.

> TBH, taking a `TryInto<T>` as input in a function like this is an anti-pattern. It forces the API to have to represent the error even when it is not applicable, even when it doesn't make sense. It would make much more sense to take an `ObjectPath` as input, and require that the caller do the validation/conversion [if necessary] themselves.

I disagree and hence whey we've the generic API.

As we discussed before, I'm not against providing API that takes ready objects but the main API should be as easy to use as possible (#321).

Comment (zeenix) [reply zbus/src/connection/builder.rs:283]: > > because it's not clear at all _when_ this can fail or _why_.
> 
> As a user, you don't need to know. If you do, look at the generic constraints. User just needs to know that they need to pass a valid object path. If they don't, they'll get an error.

I actually have never seen the "how" being documented in "Error" sections of fallible methods in the Rust world and I think that's the right thing to do.

**Comment by WhyNotHugo** [comment]
I don't know the error type of the other methods. I figured out the ones that I'm calling, and those were somewhat simpler to figure out. Just look at the next method:

https://github.com/dbus2/zbus/blob/6155d22ff1c52d36136488fb5d835a0d37484ae5/zbus/src/connection/builder.rs#L130-L138

The error is cast via `Into::into`, so I've no idea which implementation is called. I tried looking at the `zbus::error::Error` variants, but it's not obvious. Perhaps negative examples would work to figure out what times are returned and then document them?

As I use `zbus`, I come across bits of documentation that are missing and I have to sit down and figure out these nuances. I'm happy to send these back as PRs to save some work to the next person needing to use these function. I don't mind refining on wording so as to make the documentation as clear as possible. But I'm not signing up for fixing all the documentation for stuff that I haven't yet used. I'll be the right person to document those as I actually use them and understand their entire context.

**Comment by zeenix** [comment]
> But I'm not signing up for fixing all the documentation for stuff that I haven't yet used.

I never asked you to fix **all** the documentation. Please read my last comment again.

**Comment by zeenix** [comment]
> I'll be the right person to document those as I actually use them and understand their entire context.

You could at least adjust the wording for `serve_at` docs to include the `iface`?

File: zbus/src/connection/builder.rs (hunk 4 above)
Comment (MaxVerevkin) [reply zbus/src/connection/builder.rs:283]: > User just needs to know that they need to pass a valid object path.

Yeah, and IMHO it can be best described by the type system, by requiring the argument to be of a certain concrete type, not the documentation.

> I'm not against providing API that takes ready objects

I thought #321 was rejected... But good to know. Anyway, this is getting off-topic :)

Comment (zeenix) [reply zbus/src/connection/builder.rs:283]: > > User just needs to know that they need to pass a valid object path.
> 
> Yeah, and IMHO it can be best described by the type system, 

It is! The generics say what's the expected type.

> by requiring the argument to be of a certain concrete type, not the documentation.

I'm not the one pushing for docs here but I'm not against them either. We're enforcing compliance whether user has to create the types or we do it for them.

> > I'm not against providing API that takes ready objects
>
> I thought #321 was rejected... But good to know. Anyway, this is getting off-topic :)

Hmm.. actually it was. I confused with #403, where the suggestion is to add additional API.

Comment (MaxVerevkin) [reply zbus/src/connection/builder.rs:283]: > where the suggestion is to add additional API.

#321 also suggests to _add_ API (it's in the title :smile:), but #403 is _very_ similar but better written, so I'm happy to see that my idea is not dead.

**Comment by WhyNotHugo** [comment]
> You could at least adjust the wording for serve_at docs to include the iface?

Done, I think that it should be good now :)

File: zbus/src/connection/builder.rs [hunk 5]
```diff
@@ -260,12 +270,16 @@ impl<'a> Builder<'a> {
         self
     }
 
-    /// Register a D-Bus [`Interface`] to be served at a given path.
+    /// Register a D-Bus [`Interface`] `iface` to be served at a given path.
```
Comment (zeenix) [inline zbus/src/connection/builder.rs:273]: Actually this is not what I meant. I would never ask you to add unrelated changes to a commit. :) 

I was mistakenly thinking the `iface` attribute is also using `TryFrom` and is therefore fallible. Hence why I asked for it to be also included in the error docs you're adding. 

So I'm sorry for misguiding you but please remove this change.

Comment (WhyNotHugo) [reply zbus/src/connection/builder.rs]: Oh, I was a bit confused about this bit, but it has been a few days too :sweat_smile: 

Should be okay now?

Comment (zeenix) [reply zbus/src/connection/builder.rs]: @WhyNotHugo yes, thanks. :)

**Comment by zeenix** [comment]
> Looks good otherwise but apart from the mistake I made, please add these docs to the blocking wrapper too: `blocking::connection::Builder`.

This is still missing. Just need C&P.

On a related note, it would be nice to make the blocking wrapper methods' docs just point to the underlying async methods' docs so we don't need all this repetition.

**Comment by WhyNotHugo** [comment]
> On a related note, it would be nice to make the blocking wrapper methods' docs just point to the underlying async methods' docs so we don't need all this repetition.

I can add something like this to the blocking wrapper instead:

```
/// Wrapper around `connection::Builder`.
///
/// See its documentation for details (including error conditions).
```

Or I can just copy-paste it, lemme know what you prefer.

**Comment by zeenix** [comment]
> /// Wrapper around `connection::Builder`.
> ///
> /// See its documentation for details (including error conditions).

We need to do a more consistent overall change for this one. If you want to do this, please go ahead but then each method needs to be something like this:

```rust
/// Blocking version of [`connection::Builder::method_name`].
```

as a separate commit that comes before the current one in this PR.

> Or I can just copy-paste it, lemme know what you prefer.

That's fine too. You choose your pain. :)

**Comment by WhyNotHugo** [comment]
This is too much work. Honestly, I'd have preferred if you just rejected my PR and told me to fuck off.

**Comment by zeenix** [comment]
> This is too much work. Honestly, I'd have preferred if you just rejected my PR and told me to fuck off.

I don't know how you arrived at this after I agreed to just c&p:

> Or I can just copy-paste it, lemme know what you prefer.

I suggested that myself.
